### PR TITLE
[WIP] Add custom data type support

### PR DESCRIPTION
**Pull request by tigrannajaryan**
This will allow to have pipelines of data types that are not one of the
builtin type: traces and metrics.

See design.md for detailed description of how custom data types work.

### CODE REVIEW

File: service/builder/exporters_builder.go [hunk 1]
```diff
@@ -332,3 +351,22 @@ func createMetricsExporter(factoryBase component.ExporterFactoryBase,
...
+	if factory, ok := factoryBase.(component.DataExporterFactory); ok {
+		creationParams := component.ExporterCreateParams{Logger: logger}
+		ctx := context.Background()
+
+		// If exporter is of the new type (can manipulate on internal data structure),
+		// use ExporterFactory.CreateDataExporter.
+		return factory.CreateExporter(ctx, creationParams, dataType, cfg)
+	}
+
+	return nil, fmt.Errorf("exporter %q does not support data type %q", factoryBase.Type(), dataType)
```
Comment (dmitryax) [inline service/builder/exporters_builder.go:371]: I don't think this function really needed. For metrics and traces, it supposed to be a temporary function to support old->new style data model migration. What if we don't have `if factory, ok := factoryBase.(component.DataExporterFactory); ok {` condition? Looks like we would get a compile time error instead of this one which seems to be more useful.

File: service/builder/pipelines_builder.go [hunk 2]
```diff
@@ -318,3 +352,30 @@ func createMetricsProcessor(
...
+		if dataConsumer, ok := nextConsumer.(consumer.DataConsumer); ok {
+			return factory.CreateProcessor(ctx, creationParams, dataType, cfg, dataConsumer)
+		}
+
+		return nil, fmt.Errorf("processor %q is attached to a pipeline with a component that does accept data type %q",
+			cfg.Name(), dataType)
+	}
+
+	return nil, fmt.Errorf("processor %q does support data type %q",
+		cfg.Name(), dataType)
```
Comment (dmitryax) [inline service/builder/pipelines_builder.go:380]: the same question here

**Comment by owais** [comment]
This makes sense and would allow people to support a lot of other use-cases. +1 for adding something like this. I was thinking of doing this a bit differently. What if the core repo did not deal with building exporters, processors and receivers at all and only built pipeline objects? 

1. Add a new `Pipeline` component interface
2. Service parses config, iterates over pipelines in config and calls pipeline factory to build each pipeline. 
3. Pipeline's build method or pipeline factory would build exporters, processors and receivers. We'd move most of the code from builder package to pipeline. 

Service would then just interact with pipeline objects eliminating the need for service to know and deal data types handled by each pipeline. It can basically just call `Build(config, factories)` `Start(ctx, host)` and `Shutdown()` methods on each pipeline. Pipeline would then internally call these methods as needed on each of it's components. Most of the builder package would become the default pipeline component. 

Once we have this, anyone should be able to implement a different pipeline that deals with a completely different data-type and use it with Otel just like people build other custom components now. 

Implementing a custom pipeline would mean duplicating/copying a bit of code but the flexibility this will bring to the system would be worth it IMO. Also, it should be rare for people to implement custom pipelines so copying a bit of code (if required) shouldn't make that much of a difference. We can also share a lot of code by using common/base interfaces to pass around components. Pipeline might do some type-assertion on it's components but it would happen only once as part of the startup process. 

This will also allow people to customize other pipeline behavior like replacing fanout connector with something else or handling errors in a different way without a processor, etc.

In case it becomes tedious to implement a pipeline, it should be trivial to support automate it with code gen. With code gen, I'm sure we can make writing a custom pipeline as simple as the following:

```golang
// data_type.go
package myCustomPipeline

type MyCustomDataType struct {
    Name string
}

// gen.go
package myCustomPipeline

//go:generate otel-pipeline-generator -type MyCustomDataType -Name MyCustomPipeline
```

I'm sure this should be all that is required to build a pipeline with default behaviour that processes custom data types.

TL;DR: Add only one new type/component called Pipeline, re-work builder package into a default Pipeline component. Let people register custom pipelines. Add a code generator to trivialize building custom pipelines.

**Comment by jrcamp** [comment]
I think this proposal originally came out of how do we send metadata from receivers. Later we also discussed having metadata be its own formal data type alongside metrics, traces, logs. I think that will be important when deployed in collector mode and users want to be able to take metadata from one collector instance inside the firewall, forward to one outside the firewall, and then send to cloud. If it's just part of the OT protocol then it can use existing receivers/exporters.

**Comment by tigrannajaryan** [comment]
Decided to go a different route: https://github.com/open-telemetry/opentelemetry-collector/pull/959
Closing this.
